lib/Doctrine/Enumeration/Enumerated.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Enumeration;
+
+use Doctrine\Enumeration\Exception\NotSupportedException;
+use function array_search;
+use function constant;
+use function in_array;
+
+trait Enumerated
+{
+    final private function __construct()
+    {
+    }
+
+    /**
+     * @param mixed $value
+     */
+    final public static function validateValue($value) : void
+    {
+        if (! in_array($value, self::availableValues(), true)) {
+            throw new NotSupportedException('Invalid value.');
+        }
+    }
+
+    final public static function validateName(string $name) : void
+    {
+        if (! isset(self::availableValues()[$name])) {
+            throw new NotSupportedException('Invalid name.');
+        }
+    }
+
+    /**
+     * @param mixed $value
+     */
+    final public static function get($value) : self
+    {
+        self::validateValue($value);
+
+        $name      = array_search($value, self::availableValues(), true);
+        $instances = &self::registry(static::class);
```

Comment:
Maybe instead of using a method which returns a variable by reference you could use the same static class variable directly to save method calls and avoid the code like `if (!isset($registry[$type]))` in `registry()`:
```php
return self::$instances[static::class][$name] ?? self::$instances[static::class][$name] = new static();
```

Reply:
👍 cached instance should be the happy path

Reply:
The only purpose of the `&registry()` method is to hide the instances from reflection. Of course without it it's easier and likely faster, but allows anyone mess with those.

Reply:
If someone cannot hack something with reflection, they still can use stream filter applied to the class loader. There's no technical way to prevent people from being able to shoot themselves in the foot.